example/extensions/lib_subgraph/subgraph_lib.cc

```diff
@@ -0,0 +1,278 @@
+
+  // convert json string to json object
+  JsonParser parser;
+  JsonVal json_val = parser.parse_to_json(subgraph_sym);
+  // get nodes list
+  JsonVal nodes = json_val.map[JsonVal("nodes")];
+  //counter for inputs
+  int input_cnt = 0;
+  // temporary tensor storage
+  std::vector<MXTensor> data;
+  // track memory allocations to free later
+  std::vector<void*> to_free;
+
+  // loop over nodes
+  for(int i=0; i<nodes.list.size(); i++) {
+    JsonVal node = nodes.list[i];
+    // get the op name
+    std::string op = node.map[JsonVal("op")].str;
+    // get node ID inputs to op
+    JsonVal node_inputs = node.map[JsonVal("inputs")];
+    
+    // handle each op type
+    if (op.compare("null") == 0) {
+      // null is an input data to the subgraph, add to data storage
+      data.push_back(inputs[input_cnt++]);
+    } else if (op.compare("log") == 0) {
+      // get input tensor based on node ID inputs from data storage
+      MXTensor &input = data[node_inputs.list[0].list[0].num];
+      // create temporary storage
+      MXTensor tmp(malloc(input.size()*4), input.shape, input.dtype);
+      // save allocated ptr to free later
+      to_free.push_back(tmp.data_ptr);
+      // execute log operator
+      myLog(input,tmp);
+      // add output tensor to data storage
+      data.push_back(tmp);
+    } else if (op.compare("exp") == 0) {
+      // get input tensor based on node ID inputs from data storage
+      MXTensor &input = data[node_inputs.list[0].list[0].num];
+      // create temporary storage
+      MXTensor tmp(malloc(input.size()*4), input.shape, input.dtype);
+      // save allocated ptr to free later
+      to_free.push_back(tmp.data_ptr);
+      // execute exp operator 
+      myExp(input,tmp);
+      // add output tensor to data storage
+      data.push_back(tmp);
+    } else {
+      std::cout << "Error! Unsupported op '" << op << "' found in myExecutor";
+      // free allocated temporary storage
+      for (void* ptr : to_free)
+        free(ptr);
+      return MX_FAIL;
+    }
+  }
+  
+  // get list of outputs from subgraph
+  JsonVal heads = json_val.map[JsonVal("heads")];
+  // copy all operator results to outputs of subgraph
+  for (int j = 0; j < heads.list.size(); j++) {
+    // get computed result
+    MXTensor &result = data[heads.list[0].list[0].num];
+    // get output tensor to pass to MX
+    MXTensor &out = outputs[j];
+    float *out_data = out.data<float>();
+    float *res_data = result.data<float>();
+    // loop and copy data
+    for (int64_t i = 0; i < result.size(); i++) {
+      out_data[i] = res_data[i];
+    }
+  }
+
+  // free allocated temporary storage
+  for (void* ptr : to_free) {
+    free(ptr);
+  }
+  
+  return MX_SUCCESS;
+}
+
+class MyStatefulOp : public CustomStatefulOp {
+ public:
+  explicit MyStatefulOp(std::string sym) : subgraph_sym(sym) {}
+
+  MXReturnValue Forward(std::vector<MXTensor> inputs,
+                        std::vector<MXTensor> outputs,
```

Comment:
For inputs and outputs, is passing by reference better?

Reply:
Sure, please suggest here #17006

Reply:
correct, that is something we can improve for our custom xx stack down the line, so you can suggest it in that RFC